storage_controller/src/service.rs

```diff
@@ -5783,20 +5793,38 @@ impl Service {
             ));
         }
 
+        let base_node_url = node::build_base_url(
+            &register_req.listen_http_addr,
+            register_req.listen_http_port,
+            register_req.listen_https_port,
+            self.config.use_https_pageserver_api,
+        );
+        let base_node_url = match base_node_url {
+            Ok(url) => url,
+            Err(message) => return Err(ApiError::PreconditionFailed(message.into())),
+        };
+
         // Ordering: we must persist the new node _before_ adding it to in-memory state.
         // This ensures that before we use it for anything or expose it via any external
         // API, it is guaranteed to be available after a restart.
         let new_node = Node::new(
             register_req.node_id,
             register_req.listen_http_addr,
             register_req.listen_http_port,
+            register_req.listen_https_port,
             register_req.listen_pg_addr,
             register_req.listen_pg_port,
             register_req.availability_zone_id.clone(),
+            base_node_url,
         );
 
-        // TODO: idempotency if the node already exists in the database
-        self.persistence.insert_node(&new_node).await?;
+        if registration_status == RegistrationStatus::New {
+            self.persistence.insert_node(&new_node).await?;
+        } else {
+            self.persistence
+                .update_node_on_registration(register_req.node_id, register_req.listen_https_port)
+                .await?;
+        }
```

Comment:
nit: Let's match explicitly here on the `registration_status`. Guards against surprises for when we add a new enum variant. Same below.